src/main/java/org/sagebionetworks/web/client/widget/ReactComponentV2.java

```diff
@@ -0,0 +1,259 @@
+package org.sagebionetworks.web.client.widget;
+
+import com.google.gwt.core.client.GWT;
+import com.google.gwt.dom.client.DivElement;
+import com.google.gwt.dom.client.Document;
+import com.google.gwt.dom.client.Element;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.dom.client.HasClickHandlers;
+import com.google.gwt.event.shared.HandlerRegistration;
+import com.google.gwt.user.client.ui.ComplexPanel;
+import com.google.gwt.user.client.ui.Widget;
+import java.util.ArrayList;
+import java.util.List;
+import jsinterop.base.JsPropertyMap;
+import org.sagebionetworks.web.client.jsinterop.React;
+import org.sagebionetworks.web.client.jsinterop.ReactComponentProps;
+import org.sagebionetworks.web.client.jsinterop.ReactComponentType;
+import org.sagebionetworks.web.client.jsinterop.ReactDOM;
+import org.sagebionetworks.web.client.jsinterop.ReactDOMRoot;
+import org.sagebionetworks.web.client.jsinterop.ReactElement;
+
+/**
+ * Abstract widget that manages the lifecycle of a {@link React} component tree mounted with {@link ReactDOM}.
+ * <p>
+ * To use it, extend the class, supply a {@link ReactComponentType} and {@link ReactComponentProps} to the constructor,
+ * and call {@link #render()} to render the React component.
+ * <p>
+ * This widget also manages appending child elements if the associated React component can contain children. If all
+ * child widgets implement this class, then the child {@link ReactElement}s will be passed as children to the React
+ * component. If any child of this component is a non-ReactComponent widget, then all child widgets (including
+ * implementations of {@link ReactComponentV2}) will be injected into the node found using the component's `ref`.
+ * <p>
+ * The root element defaults to a `div`, but can be changed (e.g. to a `span`) using the
+ * {@link #ReactComponentV2(T, P, String)} constructor.
+ * <p>
+ * This widget automatically unmounts the ReactComponent (if any) when this container is detached/unloaded.
+ */
+public abstract class ReactComponentV2<
```

Comment:
New _abstract_ class -- override it like in the new `Grid` widget to use it.

I cleaned up the prior implementation and added some critical new behavior to fix a few bugs.